### PR TITLE
Implement Flexible Symbolication

### PR DESCRIPTION
**Pull request by mitsuhiko**
This is the work in progress pull request for #3352.  It adds flexible symbolication
for frames which allows mixing of platforms and more.  Instead of using the apple
crash report format we send our own stuff along.
## Flexible Symbolication Overview

Essentially unlike the old crash report system this does symbolication on a frame by frame basis. It requires that the clients send the data in the format that sentry expects in general as there is no service side data munching going on.  The stacktraces are sent as normal however there is per-frame support for symbol resolution.

This is implemented by combining two interfaces:
- `debug_meta`. This is a new ephemeral interface which holds metadata for resolving symbols and other things. After processing this data is discarded. This is where debug images and SDK information is stored.
- The frame keys `instruction_addr`, `symbol_addr` and `image_addr` (the latter is basically what `object_addr` is in apple's crash format). We accept addresses as integers or hexadecimal strings (with `0x` prefix) to avoid issues with JSON limitations.
### Debug Mapping Example

The debug image mapping and sdk info in the debug meta looks like this:

``` json
{
  "debug_meta": {
    "images": [
      {
        "type": "apple",
        "cpu_subtype": 0, 
        "uuid": "C05B4DDD-69A7-3840-A649-32180D341587", 
        "image_vmaddr": 4294967296, 
        "image_addr": 4295098368, 
        "cpu_type": 16777228, 
        "image_size": 32768, 
        "name": "/var/containers/Bundle/Application/B33C37A8-F933-4B6B-9FFA-152282BFDF13/SentryTest.app/SentryTest"
      }
    ],
    "sdk_info": {
      "dsym_type": "macho",
      "sdk_name": "iOS",
      "version_major": 9,
      "version_minor": 3,
      "version_patchlevel": 0
    }
  }
}
```

As you can see it mostly follows what we do with the apple crash reports already but there are some subtle differences:
1. the type of the image needs to be defined. Currently only `apple` is supported. Depending on the image type the method could change. In particular this will become relevant as we add PDB support.
2. the uuid and image address is required
3. the SDK info is needed to look up system symbols but can be left out if unused.

I'm considering picking up the sdk info from the OS context information but this is undecided so far.
### Example normal frame

``` json
{
  "function": "<redacted>",
  "abs_path": null,
  "instruction_offset": 4,
  "package": "/usr/lib/system/libdyld.dylib",
  "filename": null,
  "symbol_addr": "0x002ac28b4",
  "lineno": null,
  "in_app": false,
  "instruction_addr": "0x002ac28b8"
}
```
### Example unsymbolicated frame

``` json
{
  "function": "main",
  "package": "FooBar",
  "instruction_addr": "0x100026330",
  "symbol_addr": "0x1000262a0",
  "image_addr": "0x100020000"
}
```

Notes for the above frame:
- The `function` and `package` keys are optional. They correspond to `symbol_name` and `object_name` in the apple crash report respectively and are only used in case the lookup fails.
- All addresses could also be sent as integers but it's discouraged due to some JSON libraries having problems with numbers > 53bit
- `image_addr` is what `object_addr` is in apple crash reports.

---

This change is [Reviewable](https://reviewable.io/reviews/getsentry/sentry/3634)

### CODE REVIEW

**Comment by mitsuhiko** [comment]
The idea is that there will be tests for this. But in particular this is how things work for now:

* We have a new ephemeral interface called `debug_images` which has two keys: `images` which is a list of images which look very similar to what we have in the apple crash reports (it's called `binary_images`) there and `sdk_info` which is the SDK info we had previously internal only.
* If a frame adds `package_addr` (basically what `image_addr` is in the apple crash reports) and `instruction_addr` we will start symbolication.

Looking at this a bit more i probably want to rename `package_addr` back to `image_addr` since we look that up in the debug images so that just makes more sense.

**Comment by mitsuhiko** [comment]
I'm generally happy to merge this but I want to see if I can add some tests for this.

File: src/sentry/interfaces/debug_images.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def imagetype(name):
+    def decorator(f):
+        image_types[name] = f
+        return f
+    return decorator
+
+
+@imagetype('apple')
```
Comment (dcramer) [inline src/sentry/interfaces/debug_images.py]: if we have control over this, would it make sense to call it cocoa?

**Comment by dcramer** [comment]
I didn't go over this into too much detail, but might be good to get @tkaemming to give it a once over. +1 on the tests since this code is getting bulkier and its hard to follow some logic (and will be easy to break).

**Comment by mitsuhiko** [comment]
Theres are in and some improvements on symsynd as well. This is ready for review.

**Comment by dcramer** [comment]
There's a few things that could use some test coverage (just to ensure we dont break simple things), but I think in general it's ok. I didn't do in depth since its too big.
